go/pkg/tree/tree_test.go

```diff
@@ -0,0 +1,522 @@
+				{Name: "barDir", Digest: barDirDgPb},
+				{Name: "fooDir", Digest: fooDirDgPb},
+			}},
+			additionalBlobs: [][]byte{fooBlob, barBlob, fooDirBlob, barDirBlob},
+		},
+		{
+			// NOTE: The use of maps in our implementation means that the traversal order is unstable. The
+			// outputs are required to be in lexicographic order, so if ComputeMerkleTree is not sorting
+			// correctly, this test will fail (except in the rare occasion the traversal order is
+			// coincidentally correct).
+			desc: "Correct sorting",
+			input: []*inputPath{
+				{path: "a", fileContents: fooBlob},
+				{path: "b", fileContents: fooBlob},
+				{path: "c", fileContents: fooBlob},
+				{path: "d", fileContents: fooBlob},
+				{path: "e", fileContents: fooBlob},
+				{path: "f", fileContents: fooBlob},
+				{path: "g/foo", fileContents: fooBlob},
+				{path: "h/foo", fileContents: fooBlob},
+				{path: "i/foo", fileContents: fooBlob},
+				{path: "j/foo", fileContents: fooBlob},
+				{path: "k/foo", fileContents: fooBlob},
+				{path: "l/foo", fileContents: fooBlob},
+			},
+			spec: &command.InputSpec{
+				Inputs: []string{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l"},
+			},
+			rootDir: &repb.Directory{
+				Files: []*repb.FileNode{
+					{Name: "a", Digest: fooDgPb, IsExecutable: true},
+					{Name: "b", Digest: fooDgPb, IsExecutable: true},
+					{Name: "c", Digest: fooDgPb, IsExecutable: true},
+					{Name: "d", Digest: fooDgPb, IsExecutable: true},
+					{Name: "e", Digest: fooDgPb, IsExecutable: true},
+					{Name: "f", Digest: fooDgPb, IsExecutable: true},
+				},
+				Directories: []*repb.DirectoryNode{
+					{Name: "g", Digest: fooDirDgPb},
+					{Name: "h", Digest: fooDirDgPb},
+					{Name: "i", Digest: fooDirDgPb},
+					{Name: "j", Digest: fooDirDgPb},
+					{Name: "k", Digest: fooDirDgPb},
+					{Name: "l", Digest: fooDirDgPb},
+				},
+			},
+			additionalBlobs: [][]byte{fooBlob, fooDirBlob},
+		},
+	}
+
+	for _, tc := range tests {
+		root, err := ioutil.TempDir("", tc.desc)
+		if err != nil {
+			t.Fatalf("failed to make temp dir: %v", err)
+		}
+		defer os.RemoveAll(root)
+		if err := construct(root, tc.input); err != nil {
+			t.Fatalf("failed to construct input dir structure: %v", err)
+		}
+
+		t.Run(tc.desc, func(t *testing.T) {
+			wantBlobs := make(map[digest.Digest][]byte)
+			rootBlob := mustMarshal(t, tc.rootDir)
+			rootDg := digest.NewFromBlob(rootBlob)
+			wantBlobs[rootDg] = rootBlob
+			for _, b := range tc.additionalBlobs {
+				wantBlobs[digest.NewFromBlob(b)] = b
+			}
+
+			gotBlobs := make(map[digest.Digest][]byte)
+			gotRootDg, inputs, err := ComputeMerkleTree(root, tc.spec, chunker.DefaultChunkSize)
+			if err != nil {
+				t.Errorf("ComputeMerkleTree(...) = gave error %v, want success", err)
+			}
+			for _, ch := range inputs {
+				blob, err := ch.FullData()
+				if err != nil {
+					t.Errorf("chunker %v FullData() returned error %v", ch, err)
```

Comment:
What does this print?

Reply:
The chunker has a String() method. This prints something like:
chunker <3 bytes>: /tmp/Directory exclusions884289053/barDir/bar FullData() returned error <...>

Reply:
Thanks for clarifying!